File: cli/lib/capability-manifest.mjs
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

export const CAPABILITY_MANIFEST_PATH = fileURLToPath(
  new URL("./capability-manifest.json", import.meta.url),
);
const EXPECTED_COUNTS = {
  routes: 23,
  "install.skills": 20,
  "install.commands": 16,
  "install.agents": 4,
  "mcp.tools": 15,
  "mcp.learningProfileWriteTools": 3,
  "sdk.exports": 9,
  "sdk.learnMethods": 3,
};

function assertStringArray(value, field) {
  if (!Array.isArray(value) || value.some((item) => typeof item !== "string" || !item.trim())) {
    throw new Error(`Capability manifest ${field} must be an array of non-empty strings`);
  }
  if (new Set(value).size !== value.length) {
    throw new Error(`Capability manifest ${field} must not contain duplicates`);
  }
  if (value.length !== EXPECTED_COUNTS[field]) {
    throw new Error(
      `Capability manifest ${field} must contain ${EXPECTED_COUNTS[field]} entries, got ${value.length}`,
    );
  }
}

function assertExactKeys(value, expectedKeys, field) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`Capability manifest ${field} must be an object`);
  }
  const actualKeys = Object.keys(value);
  if (actualKeys.length !== expectedKeys.length || expectedKeys.some((key) => !Object.hasOwn(value, key))) {
    throw new Error(`Capability manifest ${field} keys must be: ${expectedKeys.join(", ")}`);
  }
}

export function validateCapabilityManifest(manifest) {
  if (!manifest || typeof manifest !== "object" || Array.isArray(manifest)) {
    throw new Error("Capability manifest must be a JSON object");
  }
  if (manifest.schemaVersion !== 1) {
    throw new Error("Capability manifest schemaVersion must be 1");
  }

  assertExactKeys(manifest, ["schemaVersion", "routes", "install", "mcp", "sdk"], "root");
  assertExactKeys(manifest.install, ["skills", "commands", "agents"], "install");
  assertExactKeys(manifest.mcp, ["tools", "learningProfileWriteTools"], "mcp");
  assertExactKeys(manifest.sdk, ["exports", "learnMethods"], "sdk");

  assertStringArray(manifest.routes, "routes");
  for (const kind of ["skills", "commands", "agents"]) {
    assertStringArray(manifest.install?.[kind], `install.${kind}`);
  }
  assertStringArray(manifest.mcp?.tools, "mcp.tools");
  assertStringArray(manifest.mcp?.learningProfileWriteTools, "mcp.learningProfileWriteTools");
  assertStringArray(manifest.sdk?.exports, "sdk.exports");
  assertStringArray(manifest.sdk?.learnMethods, "sdk.learnMethods");

  const mcpTools = new Set(manifest.mcp.tools);
  for (const toolName of manifest.mcp.learningProfileWriteTools) {
    if (!mcpTools.has(toolName)) {
      throw new Error(`Capability manifest write tool is not in mcp.tools: ${toolName}`);
    }
  }
  return manifest;
}

export function readCapabilityManifest(manifestPath = CAPABILITY_MANIFEST_PATH) {
  let manifest;
  try {
    manifest = JSON.parse(readFileSync(manifestPath, "utf8"));
  } catch (error) {
    throw new Error(`Unable to read capability manifest at ${manifestPath}: ${error.message}`);
  }
  return validateCapabilityManifest(manifest);
}

File: tools/audit/capability_manifest.py
#!/usr/bin/env python3
"""Load and validate the canonical design-ai capability identity contract."""
from __future__ import annotations

import argparse
import json
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
CAPABILITY_MANIFEST_PATH = ROOT / "cli" / "lib" / "capability-manifest.json"
EXPECTED_COUNTS = {
    "routes": 24,
    "install.skills": 21,
    "install.commands": 16,
    "install.agents": 4,
    "mcp.tools": 29,
    "mcp.learningProfileWriteTools": 3,
    "sdk.exports": 20,
    "sdk.learnMethods": 3,
}


def _validate_string_array(value: object, field: str) -> list[str]:
    if not isinstance(value, list) or any(
        not isinstance(item, str) or not item.strip()
        for item in value
    ):
        raise SystemExit(f"capability manifest {field} must be an array of non-empty strings")
    if len(set(value)) != len(value):
        raise SystemExit(f"capability manifest {field} must not contain duplicates")
    expected_count = EXPECTED_COUNTS[field]
    if len(value) != expected_count:
        raise SystemExit(
            f"capability manifest {field} must contain {expected_count} entries, got {len(value)}"
        )
    return value


def validate_capability_manifest(manifest: object) -> dict[str, object]:
    if not isinstance(manifest, dict):
        raise SystemExit("capability manifest must be a JSON object")
    if manifest.get("schemaVersion") != 1:
        raise SystemExit("capability manifest schemaVersion must be 1")
    if set(manifest) != {"schemaVersion", "routes", "install", "mcp", "sdk"}:
        raise SystemExit("capability manifest root sections changed")

    install = manifest.get("install")
    mcp = manifest.get("mcp")
    sdk = manifest.get("sdk")
    if not isinstance(install, dict) or set(install) != {"skills", "commands", "agents"}:
        raise SystemExit("capability manifest install sections changed")
    if not isinstance(mcp, dict) or set(mcp) != {"tools", "learningProfileWriteTools"}:
        raise SystemExit("capability manifest mcp sections changed")
    if not isinstance(sdk, dict) or set(sdk) != {"exports", "learnMethods"}:
        raise SystemExit("capability manifest sdk sections changed")

    _validate_string_array(manifest.get("routes"), "routes")
    _validate_string_array(install.get("skills"), "install.skills")
    _validate_string_array(install.get("commands"), "install.commands")
    _validate_string_array(install.get("agents"), "install.agents")
    tools = _validate_string_array(mcp.get("tools"), "mcp.tools")
    write_tools = _validate_string_array(
        mcp.get("learningProfileWriteTools"),
        "mcp.learningProfileWriteTools",
    )
    _validate_string_array(sdk.get("exports"), "sdk.exports")
    _validate_string_array(sdk.get("learnMethods"), "sdk.learnMethods")
    if not set(write_tools).issubset(tools):
        raise SystemExit("capability manifest learning write tools must be listed in mcp.tools")
    return manifest


def load_capability_manifest(path: Path = CAPABILITY_MANIFEST_PATH) -> dict[str, object]:
    try:
        manifest = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as error:
        raise SystemExit(f"failed to load capability manifest: {path}") from error
    return validate_capability_manifest(manifest)


SOURCE_CAPABILITIES = load_capability_manifest()


def _expect_failure(manifest: object, expected: str) -> None:
    try:
        validate_capability_manifest(manifest)
    except SystemExit as error:
        if expected not in str(error):
            raise SystemExit(f"self-test expected {expected!r}, got {error!s}") from error
        return
    raise SystemExit(f"self-test expected capability validation failure: {expected}")


def run_self_test() -> None:
    validate_capability_manifest(json.loads(json.dumps(SOURCE_CAPABILITIES)))

    cases = []
    invalid = json.loads(json.dumps(SOURCE_CAPABILITIES))
    invalid["schemaVersion"] = 2
    cases.append((invalid, "schemaVersion must be 1"))
    invalid = json.loads(json.dumps(SOURCE_CAPABILITIES))
    invalid["unexpected"] = []
    cases.append((invalid, "root sections changed"))
    invalid = json.loads(json.dumps(SOURCE_CAPABILITIES))
    invalid["routes"][0] = "   "
    cases.append((invalid, "non-empty strings"))
    invalid = json.loads(json.dumps(SOURCE_CAPABILITIES))
    invalid["routes"][-1] = invalid["routes"][0]
    cases.append((invalid, "must not contain duplicates"))
    invalid = json.loads(json.dumps(SOURCE_CAPABILITIES))
    invalid["install"]["commands"].pop()
    cases.append((invalid, "must contain 16 entries"))
    invalid = json.loads(json.dumps(SOURCE_CAPABILITIES))
    invalid["mcp"]["learningProfileWriteTools"][-1] = "design_ai_unknown_write"
    cases.append((invalid, "must be listed in mcp.tools"))

    for manifest, expected in cases:
        _expect_failure(manifest, expected)
    print("Capability manifest self-test passed")


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--self-test", action="store_true")
    args = parser.parse_args()
    if args.self_test:
        run_self_test()
        return
    print(f"Capability manifest passed: {CAPABILITY_MANIFEST_PATH}")


if __name__ == "__main__":
    main()
